walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -29,64 +29,45 @@
 public final class FindMeetingQuery {
 
   /** 
-   * Query timeslots where all attendees must be free to attend the meeting.
+   * Query timeslots to find available time slots for attendees.
+   * Optional attendees are also considered depending on several cases:
+   * The basic functionality of optional attendees is that if one or more 
+   * time slots exists so that both mandatory and optional attendees can attend, 
+   * return those time slots. Otherwise, return the time slots that fit just 
+   * the mandatory attendees. 
+   *
    * Worst-case runtime: O(events * attendees).
   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    Collection<String> attendees = request.getAttendees();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> allAttendees = new ArrayList<String>();
+    allAttendees.addAll(mandatoryAttendees);
+    allAttendees.addAll(request.getOptionalAttendees());
+    
+    // Available time list for all attendees, including optional attendees.
+    List<TimeRange> allAttendeesAvailableTimes = getAvailableTimes(events, allAttendees, request.getDuration());
 
-    // Edge case: No attendees.
-    if (attendees.isEmpty()) {
-      return Arrays.asList(TimeRange.WHOLE_DAY);
+    // Edge case: No mandatory attendees or at least one available time slot for all attendees.
+    if (mandatoryAttendees.isEmpty() || !allAttendeesAvailableTimes.isEmpty()) {
+      return allAttendeesAvailableTimes;
     }
 
     // Edge case: Meeting duration exceeds a whole day.
     if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
```

Comment:
should this edge case get executed first?

Reply:
Definitely. Thanks! done.